// d3plus-export/es/src/dom2canvas.js
import html2canvas from "html2canvas";
import canvg from "canvg-browser";
import {select, selectAll} from "d3-selection";

import svgPresets from "./svgPresets";

var defaultOptions = {
  background: false,
  callback: function () {},
  exclude: [],
  padding: 0,
  scale: 1
};

var canvgOptions = {
  ignoreMouse: true,
  ignoreAnimation: true,
  ignoreDimensions: true,
  ignoreClear: true
};

/**
    @function parseTransform
    @desc Extracts scale, x, and y position from an elements "transform" attribute, respecting cross-browser render differences.
    @param {HTMLElement} elem The element to be analyzed.
    @private
*/
function parseTransform(elem) {

  var property = select(elem).attr("transform");
  var scale = 1, x = 0, y = 0;
  if (property) {
    scale = property.match(/scale\(([^a-z]+)\)/i);
    if (scale) { scale = parseFloat(scale[1]); }
    else { scale = 1; }
    var translate = property.match(/translate\(([^a-z]+)\)/i);
    if (translate) {
      var assign;
      (assign = translate[1]
        .replace(", ", ",")
        .replace(/([^a-z]),*\s([^a-z])/gi, "$1,$2")
        .split(",")
        .map(function (d) { return parseFloat(d) * scale; }), x = assign[0], y = assign[1]);
    }
  }
  return [scale, x, y];

}

/**
    @function dom2canvas
    @desc Renders HTML/SVG elements to a shared canvas.
    @param {HTMLElement|Object|Array} elem The element or array of elements to be rendered to a single canvas. Additionally, a complex object can be passed as an element which can contain specific other properties.
    @param {Number} [elem.x] The x offset of the element within the rendered canvas.
    @param {Number} [elem.y] The y offset of the element within the rendered canvas.
    @param {Object} [options] Additional options to specify.
    @param {String} [options.background] Background color of the rendered canvas.
    @param {Function} [options.callback] Callback function to be passed the canvas element after rendering.
    @param {Array} [options.excludes] An array of HTMLElement objects to be excluded from the render.
    @param {Number} [options.height] Pixel height for the final output. If a height value has not been passed, it will be inferred from the sizing of the first DOM element passed.
    @param {Number} [options.padding = 0] Outer padding for the final file.
    @param {Number} [options.scale = 1] Scale for the final file.
    @param {Number} [options.width] Pixel width for the final output. If a width value has not been passed, it will be inferred from the sizing of the first DOM element passed.
*/
export default function(elem, options) {

  if (!elem) { return; }
  if (!(elem instanceof Array)) { elem = [elem]; }

  options = Object.assign({}, defaultOptions, options);
  var IE = new RegExp(/(MSIE|Trident\/|Edge\/)/i).test(navigator.userAgent);
  var ratio = window ? window.devicePixelRatio || 1 : 1;

  var reference = elem[0];
  if (reference.constructor === Object) { reference = reference.element; }

  var height = options.height || parseFloat(select(reference).style("height")),
        width = options.width || parseFloat(select(reference).style("width"));

  var layerX, layerY, offsetX = 0, offsetY = 0;
  if (reference.getBoundingClientRect) {
    var bounds = reference.getBoundingClientRect();
    offsetX = bounds.left;
    offsetY = bounds.top;
  }
  else {
    offsetX = reference.offsetLeft;
    offsetY = reference.offsetTop;
  }

  var canvas = document.createElement("canvas");
  canvas.width = (width + options.padding * 2) * options.scale * ratio;
  canvas.height = (height + options.padding * 2) * options.scale * ratio;
  canvas.style.width = (width + options.padding * 2) * options.scale;
  canvas.style.height = (height + options.padding * 2) * options.scale;

  var context = canvas.getContext("2d");
  context.scale(options.scale * ratio, options.scale * ratio);
  context.clearRect(0, 0, canvas.width / 2, canvas.height / 2);

  if (options.background) {
    context.beginPath();
    context.rect(0, 0, canvas.width / 2, canvas.height / 2);
    context.fillStyle = options.background;
    context.fill();
  }

  var layers = [];

  function checkRender(trans) {

    var tag = (this.tagName || "").toLowerCase();
    if (options.exclude.includes(this) || tag === "foreignobject") { return; }

    var transform = Object.assign({}, trans);

    // strips translate and scale from transform property
    if (this.tagName) {

      var opacity = select(this).attr("opacity") || select(this).style("opacity");
      var display = select(this).style("display");
      var visibility = select(this).style("visibility");
      if (display === "none" || visibility === "hidden" || opacity && parseFloat(opacity) === 0) { return; }

      var tag$1 = this.tagName.toLowerCase();
      var ref = parseTransform(this);
      var scale = ref[0];
      var x = ref[1];
      var y = ref[2];

      if (tag$1 === "g") {
        transform.scale *= scale;
        transform.x += x;
        transform.y += y;
      }

      if (tag$1 === "svg") {
        var rect = this.getBoundingClientRect();
        transform.x += rect.left - offsetX;
        transform.y += rect.top - offsetY;

        var x$1 = select(this).attr("x");
        x$1 = x$1 ? parseFloat(x$1) * transform.scale : 0;
        transform.x += x$1;
        var y$1 = select(this).attr("y");
        y$1 = y$1 ? parseFloat(y$1) * transform.scale : 0;
        transform.y += y$1;
        transform.clip = {
          height: parseFloat(select(this).attr("height") || select(this).style("height")),
          width: parseFloat(select(this).attr("width") || select(this).style("width")),
          x: x$1, y: y$1
        };
      }
      else {
        var x$2 = select(this).attr("x");
        if (x$2) { transform.x += parseFloat(x$2) * transform.scale; }
        var y$2 = select(this).attr("y");
        if (y$2) { transform.y += parseFloat(y$2) * transform.scale; }
      }

    }

    if (!tag.length) {
      var test = (this.wholeText || "").replace(/\s/g, "");
      if (test.length) {

        var text = this.nodeValue
          .replace(/^\s*/, "")
          .replace(/^\n/, "")
          .replace(/^\s*/, "")
          .replace(/\n$/, "")
          .replace(/\s*$/, "")
          .replace(/\n$/, "");

        layers.push({type: "text", style: this.parentNode, value: text, x: transform.x, y: transform.y});

      }
    }
    else if (tag === "defs") { return; }
    else if (tag === "text") {
      var elem = this.cloneNode(true);
      select(elem).call(svgPresets);
      layers.push(Object.assign({}, transform, {type: "svg", value: elem}));
    }
    else if (["image", "img"].includes(tag)) {

      var url = select(this).attr("href") || select(this).attr("xlink:href");

      if (url.length) {

        var h = parseFloat(select(this).attr("height")) * transform.scale,
              w = parseFloat(select(this).attr("width")) * transform.scale;

        var data = {
          clip: transform.clip,
          height: h,
          loaded: false,
          type: "img",
          width: w,
          x: transform.x,
          y: transform.y
        };
        layers.push(data);

        var img = new Image();
        img.crossOrigin = "Anonymous";
        img.onload = function() {

          var canvas2 = document.createElement("canvas");
          var ctx2 = canvas2.getContext("2d");
          canvas2.height = h * ratio;
          canvas2.width = w * ratio;
          ctx2.drawImage(this, 0, 0, w * ratio, h * ratio);
          var himg = document.createElement("img");
          himg.src = canvas2.toDataURL("image/png");

          data.value = himg;
          data.loaded = true;

        };
        img.src = url;

      }

    }
    else if (["div", "span"].includes(tag) && !select(this).selectAll("svg").size()) {

      var data$1 = {
        height: height,
        loaded: false,
        type: "html",
        width: width,
        x: layerX - offsetX,
        y: layerY - offsetY
      };

      var tempCanvas = document.createElement("canvas");
      tempCanvas.width = (width + options.padding * 2) * options.scale * ratio;
      tempCanvas.height = (height + options.padding * 2) * options.scale * ratio;

      var tempContext = tempCanvas.getContext("2d");
      tempContext.scale(options.scale * ratio, options.scale * ratio);

      layers.push(data$1);
      html2canvas(this, {
        allowTaint: true,
        canvas: tempCanvas,
        height: height,
        width: width
      }).then(function (c) {
        data$1.value = c;
        data$1.loaded = true;
      });

    }
    else if (tag !== "svg" && this.childNodes.length > 0 && !select(this).selectAll("image, img, svg").size()) {

      var elem$1 = this.cloneNode(true);
      select(elem$1).selectAll("*").each(function() {
        select(this).call(svgPresets);
        if (select(this).attr("opacity") === "0") { this.parentNode.removeChild(this); }
      });

      layers.push(Object.assign({}, transform, {type: "svg", value: elem$1, tag: tag}));

    }
    else if (this.childNodes.length > 0) {
      checkChildren(this, transform);
    }
    else { // catches all SVG shapes

      // console.log(this);

      var elem$2 = this.cloneNode(true);
      select(elem$2).selectAll("*").each(function() {
        if (select(this).attr("opacity") === "0") { this.parentNode.removeChild(this); }
      });

      if (tag === "line") {
        select(elem$2).attr("x1", parseFloat(select(elem$2).attr("x1")) + transform.x);
        select(elem$2).attr("x2", parseFloat(select(elem$2).attr("x2")) + transform.x);
        select(elem$2).attr("y1", parseFloat(select(elem$2).attr("y1")) + transform.y);
        select(elem$2).attr("y2", parseFloat(select(elem$2).attr("y2")) + transform.y);
      }
      else if (tag === "path") {
        var ref$1 = parseTransform(elem$2);
        var scale$1 = ref$1[0];
        var x$3 = ref$1[1];
        var y$3 = ref$1[2];
        if (select(elem$2).attr("transform")) { select(elem$2).attr("transform", ("scale(" + scale$1 + ")translate(" + (x$3 + transform.x) + "," + (y$3 + transform.y) + ")")); }
      }
      select(elem$2).call(svgPresets);

      var fill = select(elem$2).attr("fill");
      var defFill = fill && fill.indexOf("url") === 0;
      // if (defFill) select(elem).attr("fill-opacity", 0);

      layers.push(Object.assign({}, transform, {type: "svg", value: elem$2, tag: tag}));
      if (defFill) {
        var def = select(fill.slice(4, -1)).node().cloneNode(true);
        var defTag = (def.tagName || "").toLowerCase();
        if (defTag === "pattern") {

          var ref$2 = parseTransform(elem$2);
          var scale$2 = ref$2[0];
          var x$4 = ref$2[1];
          var y$4 = ref$2[2];
          transform.scale *= scale$2;
          transform.x += x$4;
          transform.y += y$4;
          checkChildren(def, transform);

        }
      }

    }

  }

  function checkChildren(e, trans) {
    selectAll(e.childNodes).each(function() {
      checkRender.bind(this)(trans);
    });
  }

  for (var i = 0; i < elem.length; i++) {

    var e = elem[i],
        options$1 = {scale: 1, x: 0, y: 0};

    if (e.constructor === Object) {
      options$1 = Object.assign(options$1, e);
      e = e.element;
    }
    layerX = options$1.x;
    layerY = options$1.y;
    checkRender.bind(e)(options$1);

  }

  function checkStatus() {

    var allDone = true;
    for (var i = 0; i < layers.length; i++) {
      if (layers[i].loaded === false) {
        allDone = false;
        break;
      }
    }

    if (allDone) { finish(); }
    else { setTimeout(checkStatus, 500); }

  }

  // Wait for all images to load
  checkStatus();

  function finish() {

    for (var i = 0; i < layers.length; i++) {

      var layer = layers[i];
      var clip = layer.clip || {height: height, width: width, x: 0, y: 0};

      switch (layer.type) {

        case "img":
          context.save();
          context.beginPath();
          context.translate(options.padding + clip.x, options.padding + clip.y);
          context.rect(0, 0, clip.width, clip.height);
          context.clip();
          context.drawImage(layer.value, layer.x + clip.x, layer.y + clip.y, layer.width, layer.height);
          context.restore();
          break;

        case "html":
          context.save();
          context.beginPath();
          context.translate(options.padding, options.padding);
          context.drawImage(layer.value, layer.x, layer.y, layer.width, layer.height);
          context.restore();
          break;

        case "text":

          var parent = select(layer.style);
          var title = layer.value
            .replace(/&/g, "&amp;")
            .replace(/"/g, "&quot;")
            .replace(/'/g, "&#039;");

          var fC = parent.style("color"),
                fS = parent.style("font-size");

          var fF = parent.style("font-family").split(",")[0];

          if (fF.indexOf("'") !== 0) { fF = "'" + fF + "'"; }
          var text = "<text stroke='none' dy='" + fS + "' fill='" + fC + "' font-family=" + fF + " font-size='" + fS + "'>" + title + "</text>";

          context.save();
          context.translate(options.padding, options.padding);
          canvg(canvas, text, Object.assign({}, canvgOptions, {offsetX: layer.x, offsetY: layer.y}));
          context.restore();

          break;

        case "svg":
          var outer = IE ? (new XMLSerializer()).serializeToString(layer.value) : layer.value.outerHTML;
          context.save();
          context.translate(options.padding + clip.x, options.padding + clip.y);
          context.rect(0, 0, clip.width, clip.height);
          context.clip();
          canvg(canvas, outer, Object.assign({}, canvgOptions, {offsetX: layer.x + clip.x, offsetY: layer.y + clip.y}));
          context.restore();
          break;

        default:
          console.warn("uncaught", layer);
          break;

      }

    }

    options.callback(canvas);

  }

}
