// d3plus-shape/es/src/Shape/Path.js
import {accessor, constant} from "d3plus-common";

import Shape from "./Shape";
import largestRect from "../geom/largestRect";
import path2polygon from "../geom/path2polygon";

/**
    @class Path
    @extends Shape
    @desc Creates SVG Paths based on an array of data.
*/
var Path = (function (Shape) {
  function Path() {
    var this$1 = this;

    Shape.call(this, "path");
    this._d = accessor("path");
    this._labelBounds = function (d, i, aes) {
      var r = largestRect(aes.points, {angle: this$1._labelConfig.rotate ? this$1._labelConfig.rotate(d, i) : 0});
      return {angle: r.angle, width: r.width, height: r.height, x: r.cx - r.width / 2, y: r.cy - r.height / 2};
    };
    this._name = "Path";
    this._labelConfig = Object.assign(this._labelConfig, {
      textAnchor: "middle",
      verticalAlign: "middle"
    });
  }

  if ( Shape ) Path.__proto__ = Shape;
  Path.prototype = Object.create( Shape && Shape.prototype );
  Path.prototype.constructor = Path;

  /**
      @memberof Path
      @desc Given a specific data point and index, returns the aesthetic properties of the shape.
      @param {Object} *data point*
      @param {Number} *index*
      @private
  */
  Path.prototype._aes = function _aes (d, i) {
    return {points: path2polygon(this._d(d, i))};
  };

  /**
      @memberof Path
      @desc Draws the paths.
      @param {Function} [*callback*]
      @chainable
  */
  Path.prototype.render = function render (callback) {

    Shape.prototype.render.call(this, callback);

    this._enter
      .attr("opacity", 0)
      .attr("d", this._d)
      .call(this._applyStyle.bind(this))
      .transition(this._transition)
      .attr("opacity", 1);

    this._update.transition(this._transition)
      .call(this._applyStyle.bind(this))
      .attr("opacity", 1)
      .attr("d", this._d);

    this._exit.transition(this._transition)
      .attr("opacity", 0);

    return this;

  };

  /**
      @memberof Path
      @desc If *value* is specified, sets the "d" attribute accessor to the specified function or number and returns the current class instance.
      @param {Function|String} [*value*]
      @chainable
      @example
function(d) {
  return d.path;
}
  */
  Path.prototype.d = function d (_) {
    return arguments.length ? (this._d = typeof _ === "function" ? _ : constant(_), this) : this._d;
  };

  return Path;
}(Shape));

export default Path;
